File: trunk/tools/filter/qssign.c
/**
 * Utilities for the quality of service module mod_qos.
 *
 * Log data signing tool to ensure data integrity.
 *
 * See http://sourceforge.net/projects/mod-qos/ for further
 * details.
 *
 * Copyright (C) 2010 Pascal Buchbinder
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 *
 */

static const char revision[] = "$Id: qssign.c,v 1.1 2010-08-13 19:43:14 pbuchbinder Exp $";

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

/* openssl */
#include <openssl/evp.h>
#include <openssl/hmac.h>

/* apr/apr-util */
#include <apr_base64.h>

#define MAX_LINE 65536
#define CR 13
#define LF 10
#define SEQDIG "12"
/*
 * reads a line from stdin
 */
int qs_getLine(char *s, int n) {
  int i = 0;
  while (1) {
    s[i] = (char)getchar();
    if(s[i] == EOF) return 0;
    if (s[i] == CR) {
      s[i] = getchar();
    }
    if ((s[i] == 0x4) || (s[i] == LF) || (i == (n - 1))) {
      s[i] = '\0';
      return 1;
    }
    ++i;
  }
}

static void qs_sign(const char *sec) {
  int sec_len = strlen(sec);
  long nr = 0;
  char line[MAX_LINE];
  int dig = atoi(SEQDIG);
  int line_size = sizeof(line) - 1 - dig; /* <data> ' ' <sequence number> */
  while(qs_getLine(line, line_size)) {
    HMAC_CTX ctx;
    unsigned char data[HMAC_MAX_MD_CBLOCK];
    unsigned int len;
    char *m;
    int data_len;
    sprintf(&line[strlen(line)], " %."SEQDIG"ld", nr);
    HMAC_Init(&ctx, sec, sec_len, EVP_sha1());
    HMAC_Update(&ctx, (const unsigned char *)line, strlen(line));
    HMAC_Final(&ctx, data, &len);
    m = calloc(1, apr_base64_encode_len(len) + 1);
    data_len = apr_base64_encode(m, (char *)data, len);
    m[data_len] = '\0';
    printf("%s#%s\n", line, m);
    free(m);
    nr++;
  }
  return;
}

int main(int argc, const char * const argv[]) {
  const char sec[] = "123";
  qs_sign(sec);
  return 0;
}

File: tools/logstash-filter-qssign/lib/logstash/filters/qssign.rb
# encoding: utf-8
require "logstash/filters/base"
require "logstash/namespace"

#require 'digest'

#
# This "qssign" filter validates the signature and sequence number
# of log messages and sets the "signature" field accordingly.
#
# You may use http://opensource.adnovum.ch/mod_qos/qssign.1.html
# to sign your log data before storing/transferring it.
#
# See http://opensource.adnovum.ch/mod_qos/ for further
# details.
#
# Copyright (C) 2015 Pascal Buchbinder
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is released under the GPL with the additional
# exemption that compiling, linking, and/or using OpenSSL is allowed.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
# MA 02110-1301, USA.

class LogStash::Filters::Qssign < LogStash::Filters::Base

  config_name "qssign"
  
  # Sample configuration:
  #
  # filter {
  #   grok {
  #     match => [ "message", "%{GREEDYDATA:data} %{INT:sequence}#%{NOTSPACE:hmac}" ]
  #     tag_on_failure => [ ]
  #   }
  #   if [data] {
  #     qssign {
  #       message => "data"
  #       source => "path"
  #       sequence => "sequence"
  #       hmac => "hmac"
  #       secret => "password"
  #     }
  #     mutate {
  #       replace => [ "message", "%{data}" ]
  #       remove_field => [ "data" ]
  #     }
  #   } else {
  #     mutate {
  #       add_field => [ "signature", "missing" ]
  #     }
  #   }
  # }

  config :message, :validate => :string, :default => "data"
  config :source, :validate => :string, :default => "path"
  config :sequence, :validate => :string, :default => "sequence"
  config :hmac, :validate => :string, :default => "hmac"
  config :secret, :validate => :string, :default => "12345"

  def initialize(params)
    super(params)
    @sources = Hash.new
  end
  
  public
  def register

  end # def register

  public
  def filter(event)
    validHMAC = 0

    # read the event fields...
    # path of the file:
    source = event[@source]
    # sequence:
    sequence = event[@sequence]
    # hmac:
    hmac = event[@hmac]
    # data:
    data = event[@message]

    # configured shared secret:
    key = @secret

    # raw message:
    raw = data + " " + sequence

    # calculate hmac
    digest = OpenSSL::Digest.new('sha1')
    hmacin = OpenSSL::HMAC.digest(digest, key, raw)
    encodedHmacin = Base64.strict_encode64(hmacin)

    if (encodedHmacin <=> hmac) == 0
      if encodedHmacin.length > 0
        validHMAC = 1
      end
    end
    if validHMAC == 1
      sequenceNr = sequence.to_i
      expected = @sources[source]
      if expected == nil
        # first message from this source
        event["signature"] = "valid"
      else
        if expected == sequenceNr
          event["signature"] = "valid"
        else
          event["signature"] = "wrong sequence"
        end
      end
      nextSequence = sequenceNr + 1
      @sources[source] = nextSequence
      if data.end_with?(" qssign---end-of-data")
        @sources[source] = 1
      end
    else
      event["signature"] = "invalid"
    end
    
    filter_matched(event)
  end # def filter
end # class LogStash::Filters::Qssign
